domain/resource/state/resource.go

```diff
@@ -407,198 +421,604 @@ WHERE uuid = $resourceIdentity.uuid`,
-		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		// Check resource exists.
+		err = tx.Query(ctx, checkResourceExistsStmt, unitResourceInput).Get(&unitResourceInput)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("resource %s: %w", resourceUUID, resourceerrors.ResourceNotFound)
+		} else if err != nil {
 			return errors.Capture(err)
 		}
 
-		// Update retrieved by if it is not retrieved.
+		// Check unit exists.
+		err = tx.Query(ctx, checkValidUnitStmt, unitResourceInput).Get(&unitResourceInput)
 		if errors.Is(err, sqlair.ErrNoRows) {
-			err = tx.Query(ctx, getRetrievedByTypeStmt, retrievedTypeParam).Get(&retrievedTypeParam)
-			if errors.Is(err, sqlair.ErrNoRows) {
-				return resourceerrors.UnknownRetrievedByType
-			}
-			if err != nil {
-				return errors.Capture(err)
-			}
+			return errors.Errorf("unit %s: %w", unitUUID, resourceerrors.UnitNotFound)
+		} else if err != nil {
+			return errors.Capture(err)
+		}
 
-			// Insert retrieved by.
-			retrievedByParam.RetrievedByTypeID = retrievedTypeParam.ID
-			err = tx.Query(ctx, insertRetrievedByStmt, retrievedByParam).Run()
-			if err != nil {
-				return errors.Capture(err)
-			}
+		// Unset any existing resources with the same charm resource as the
+		// resource being set in the unit resource table.
+		err = st.unsetUnitResourcesWithSameCharmResource(ctx, tx, resourceUUID)
+		if err != nil {
+			return errors.Errorf(
+				"removing previously set unit resources for resource %s: %w", resourceUUID, err,
+			)
 		}
 
-		// update unit resource table.
+		// Update unit resource table.
 		err = tx.Query(ctx, insertUnitResourceStmt, unitResourceInput).Run()
 		return errors.Capture(err)
 	})
 
-	return resource.SetUnitResourceResult{
-		UUID:      coreresource.UUID(unitResourceInput.ResourceUUID),
-		Timestamp: unitResourceInput.AddedAt,
-	}, err
+	return err
 }
 
-// OpenApplicationResource returns the metadata for a resource.
-func (st *State) OpenApplicationResource(
-	ctx context.Context,
-	resourceUUID coreresource.UUID,
-) (resource.Resource, error) {
-	return resource.Resource{}, nil
-}
+// unsetUnitResourcesForCharmResource removes all unit resources that use a
+// charm resource.
+func (st *State) unsetUnitResourcesWithSameCharmResource(
+	ctx context.Context, tx *sqlair.TX, uuid coreresource.UUID) error {
+	resUUID := resourceUUID{UUID: uuid.String()}
+
+	// Check if there is a resource on the unit that is using the same charm
+	// resource as the resource we are trying to set. This will be an old
+	// application resource of the units' which needs to be unset.
+	checkForResourcesStmt, err := st.Prepare(`
+SELECT ur.resource_uuid AS &resourceUUID.uuid
+FROM   unit_resource ur
+JOIN   resource r ON ur.resource_uuid = r.uuid
+WHERE  (r.charm_uuid, r.charm_resource_name) IN (
+    SELECT charm_uuid, charm_resource_name
+	FROM   resource 
+	WHERE  uuid = $resourceUUID.uuid
+)`, resUUID)
```

Comment:
todo: add check for resource.state_id = 0, otherwise the method will fail at ln 1008 when the list of matching UUIDs don't. Potential resources don't get added to the unit_resource table.

Reply:
Good point, added that in.